File: scanner.rb
#!/usr/bin/env ruby

class Callback
    def initialize()
    end
    def callback(input_file, frames)
    end
    def needs_processing(input_file)
    end
end

class DirectoryCallback
    def initialize()
    end
    def callback(input_folder)
    end
    def needs_processing(input_file)
    end
end

class Scanner
    def initialize(directory, input_extension, callback_list, directory_callback_list)
        @directory = File.expand_path(directory)
        @input_extension = input_extension
        @callback_list = Array.new
        @callback_list.push(callback_list)
        @callback_list.flatten!
        @directory_callback_list = Array.new
        @directory_callback_list.push(directory_callback_list)
        @directory_callback_list.flatten!
        throw "ffmpeg is required" if !is_ffmpeg_present()
    end

    def is_ffmpeg_present()
        result = `which ffmpeg`
        return false if nil == result || result.empty?
        return true
    end
    private :is_ffmpeg_present

    def timecode_string(seconds)
        hours = (seconds / (60 * 60)).to_i
        seconds = seconds % (60 * 60)
        minutes = (seconds / 60).to_i
        seconds = seconds % 60
        return sprintf("%02u:%02u:%02u.000", hours, minutes, seconds)
    end
    private :timecode_string

    def extract_frames(video_filename, extract_period, tmp_location)
        seconds = 0
        counter = 0
        print("Extracting frames from #{video_filename}\n")
        while true
            timecode = timecode_string(seconds)
            filename = sprintf("%s/img%05d.jpg", tmp_location, counter)
            cmd = "ffmpeg -loglevel 16 -ss #{timecode} -i \"#{video_filename}\" -frames:v 1 \"#{filename}\""
            print("#{timecode}\r")
            rc = Kernel.system(cmd)
            break if (true != rc)
            break if (!File.exists?(filename))
            seconds += extract_period
            counter += 1
        end
    end
    private :extract_frames
    
    def load_frames(folder, frames)
        Dir.new(folder).each { |filename|
            if (0 == filename.index('img') && filename.end_with?('.jpg'))
                frames << File.expand_path(folder + '/' + filename);
            end
        }
    end
    private :load_frames
    
    def scan()
        # First process all files recursively.
        scan_dir_for_files(@directory)
        # After all of the artifacts have been generated, scan all folders recursively.
        scan_dir_for_folders(@directory)
    end

    def scan_dir_for_files(directory_name)
        Dir.open(directory_name) { |dir|
            file_list = []
            dir.each { |item|
                next if item.empty? || '.' == item[0]
                full_path = File.expand_path(directory_name + "/" + item)
                
                if File.directory?(full_path)
                    scan_dir_for_files(full_path) 
                    next
                end
                next if !full_path.end_with?(@input_extension)
                
                any_need_processing = false
                @callback_list.each { |obj|
                    any_need_processing = any_need_processing | obj.needs_processing(full_path)
                }
                
                if any_need_processing
                    file_list << full_path
                end
            }
            if !file_list.empty?
                file_list.sort!
                file_list.each { |full_path|
                    temp_folder = "/tmp/tagger/"
                    `rm -rf #{temp_folder}`
                    Dir.mkdir(temp_folder)
                    frames = Array.new
                    extract_frames(full_path, 1, temp_folder)
                    load_frames(temp_folder, frames)
                    @callback_list.each { |obj|
                        obj.callback(full_path, frames)
                    }
                }
            end
        }
    end
    private :scan_dir_for_files

    def scan_dir_for_folders(directory_name)
        directory_name = File.expand_path(directory_name)
        Dir.open(directory_name) { |dir|
            dir.each { |item|
                next if item.empty? || '.' == item[0]
                full_path = File.expand_path(directory_name + "/" + item)
                
                if File.directory?(full_path)
                    # Depth-first scan.
                    scan_dir_for_folders(full_path) 
                    # Check if we need to do the thing.
                    @directory_callback_list.each { |obj|
                        if obj.needs_processing(full_path)
                            obj.callback(full_path)
                        end
                    }
                end
            }
        }
        # Finally, process the outtermost folder
        @directory_callback_list.each { |obj|
            if obj.needs_processing(directory_name)
                obj.callback(directory_name)
            end
        }
    end
    private :scan_dir_for_folders    
end

File: scanner.py
"""Core scanner engine and plugin base classes for SecurityCamProcessor."""

import glob
import os
import shutil
import subprocess


class Callback:
    """Base class for file-level processing plugins."""

    def needs_processing(self, input_file: str) -> bool:
        """Return True if this callback needs to process the given file."""
        return False

    def callback(self, input_file: str, frames: list[str]) -> None:
        """Process the given input file with the provided extracted frames."""
        pass


class DirectoryCallback:
    """Base class for directory-level processing plugins."""

    def needs_processing(self, input_dir: str) -> bool:
        """Return True if this callback needs to process the given directory."""
        return False

    def callback(self, input_dir: str) -> None:
        """Process the given directory."""
        pass


class Scanner:
    """Core engine that traverses directories, extracts frames, and dispatches to plugins."""

    directory: str
    input_extension: str
    callback_list: list[Callback]
    directory_callback_list: list[DirectoryCallback]

    def __init__(
        self,
        directory: str,
        input_extension: str,
        callback_list: list[Callback],
        directory_callback_list: list[DirectoryCallback],
    ) -> None:
        self.directory = os.path.expanduser(os.path.expandvars(os.path.abspath(directory)))
        self.input_extension = input_extension
        self.callback_list = list(callback_list)
        self.directory_callback_list = list(directory_callback_list)
        if not self._is_ffmpeg_present():
            raise RuntimeError("ffmpeg is required")

    def scan(self) -> None:
        """Run full two-phase scan: files first, then directories."""
        self._scan_dir_for_files(self.directory)
        self._scan_dir_for_folders(self.directory)

    def _is_ffmpeg_present(self) -> bool:
        """Check if ffmpeg is available on the system."""
        result: subprocess.CompletedProcess[bytes] = subprocess.run(
            ["which", "ffmpeg"], capture_output=True
        )
        return result.returncode == 0 and len(result.stdout.strip()) > 0

    def _timecode_string(self, seconds: int) -> str:
        """Convert seconds to HH:MM:SS.000 timecode format."""
        hours: int = seconds // 3600
        remainder: int = seconds % 3600
        minutes: int = remainder // 60
        secs: int = remainder % 60
        return f"{hours:02d}:{minutes:02d}:{secs:02d}.000"

    def _extract_frames(
        self, video_filename: str, extract_period: int, tmp_location: str
    ) -> None:
        """Extract frames from video at given intervals using ffmpeg."""
        seconds: int = 0
        counter: int = 0
        print(f"Extracting frames from {video_filename}")
        while True:
            timecode: str = self._timecode_string(seconds)
            filename: str = os.path.join(tmp_location, f"img{counter:05d}.jpg")
            cmd: list[str] = [
                "ffmpeg", "-loglevel", "16",
                "-ss", timecode,
                "-i", video_filename,
                "-frames:v", "1",
                filename,
            ]
            print(f"{timecode}\r", end="")
            result: subprocess.CompletedProcess[bytes] = subprocess.run(cmd, capture_output=True)
            if result.returncode != 0:
                break
            if not os.path.exists(filename):
                break
            seconds += extract_period
            counter += 1

    def _load_frames(self, folder: str) -> list[str]:
        """Read sorted img*.jpg files from the given directory."""
        pattern: str = os.path.join(folder, "img*.jpg")
        matches: list[str] = sorted(glob.glob(pattern))
        return [os.path.abspath(f) for f in matches]

    def _scan_dir_for_files(self, directory_name: str) -> None:
        """Recursively scan for matching files, extract frames, invoke callbacks."""
        try:
            entries: list[str] = os.listdir(directory_name)
        except OSError:
            return

        file_list: list[str] = []
        for item in entries:
            if not item or item.startswith("."):
                continue
            full_path: str = os.path.abspath(os.path.join(directory_name, item))
            if os.path.isdir(full_path):
                self._scan_dir_for_files(full_path)
                continue
            if not full_path.endswith(self.input_extension):
                continue
            any_need_processing: bool = False
            for cb in self.callback_list:
                if cb.needs_processing(full_path):
                    any_need_processing = True
            if any_need_processing:
                file_list.append(full_path)

        if file_list:
            file_list.sort()
            temp_folder: str = "/tmp/tagger/"
            for full_path in file_list:
                if os.path.exists(temp_folder):
                    shutil.rmtree(temp_folder)
                os.makedirs(temp_folder)
                self._extract_frames(full_path, 1, temp_folder)
                frames: list[str] = self._load_frames(temp_folder)
                for cb in self.callback_list:
                    cb.callback(full_path, frames)

    def _scan_dir_for_folders(self, directory_name: str) -> None:
        """Depth-first recursive traversal, invoke directory callbacks."""
        directory_name = os.path.abspath(directory_name)
        try:
            entries: list[str] = os.listdir(directory_name)
        except OSError:
            return

        for item in entries:
            if not item or item.startswith("."):
                continue
            full_path: str = os.path.abspath(os.path.join(directory_name, item))
            if os.path.isdir(full_path):
                self._scan_dir_for_folders(full_path)
                for dcb in self.directory_callback_list:
                    if dcb.needs_processing(full_path):
                        dcb.callback(full_path)

        # Finally, process the outermost folder
        for dcb in self.directory_callback_list:
            if dcb.needs_processing(directory_name):
                dcb.callback(directory_name)
